Try stickier block toolbars in Customize Widgets

stokesman opened the pull request:
An experiment in how to work out the recent toolbar overlapping issues in the Customize Widgets editor. The two high-level changes:
* The floating block toolbar is stickier (remains stuck at the edge of the scrolling context even when its block is out of view) which avoids a need for #32061.
* The editor header toolbar is made sticky and offsets the block toolbar (both floating and top/fixed) which will resolve #30912.

UPDATE: The floating block toolbar now hides when its block goes out of view (its position is still stuck to the scrolling boundaries to avoid overlap during transition).

## How has this been tested?
Manually…

## Screenshots

https://user-images.githubusercontent.com/9000376/122094839-76c5f880-cdc1-11eb-929b-5a9623021c3e.mp4

<details>
<summary>For posterity, the version with alternate transform transition of the toolbar</summary>

https://user-images.githubusercontent.com/9000376/121815962-c83d7e80-cc2d-11eb-9b42-6081e10caab5.mp4

</details>

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

stokesman committed:
Make sticky top of block toolbars configurable

stokesman committed:
Make the editor header sticky and have it offset the block toolbar.

stokesman committed:
Have floating block toolbar remain in view when block is out of view

- Add props to BlockTools and Popover to support expanded sticky
  area and sticky position off the bottom
- Add logic for sticking the popover as its anchorRect overflows the
  bottom of its boundary
- Remove now extraneous z-index rules and map entry

talldan commented:
@stokesman This is quite a big change (in scope and lines of code), especially for consideration as part of WordPress 5.8.

I started looking at it but found it hard to review (and I realise it's based on another PR). Would be good to see some more justification for the decisions taken and what alternatives were considered in the process of developing this.

Also would be great to see some more comments throughout the code or attempts to make the code more self-documenting. 

I found the fact that popover now has lots of different `sticky` props confusing. Popover's props feel like they need to be generally revisited. I understood what the existing 'sticky boundary' was, but now there's 'sticky' and 'sticky area' too.

The `useScrollback` hook needs documentation as this kind of mathematical code is always hard to understand for a reader.

stokesman commented:
Thanks for the feedback @talldan. I did intend to straighten it up before marking as ready for review and your comments will aid that effort.

>I found the fact that popover now has lots of different sticky props confusing.

It'd be possible to reduce flexibility a little and have fewer props. `stickyBottom` isn't actually needed for this use case and the 'sticky area' pair could be absorbed by a single boolean of `freeRangeSticky` or something. 

>Popover's props feel like they need to be generally revisited

👍

stokesman committed:
Reduce props to fit present use case

stokesman commented on packages/customize-widgets/src/components/sidebar-block-editor/index.js:
```diff
@@ -87,6 +139,24 @@ export default function SidebarBlockEditor( {
+	// toolbars are offset by the header.
+	const sectionMetaHeight = sectionMeta.offsetHeight;
+	const scrollback = useScrollback( scrollingContext, sectionMetaHeight );
+	const [ headerHeight, updateHeaderHeight ] = useReducer(
+		( height, { offsetHeight } ) => offsetHeight,
+		0
+	);
+	const headerRef = useRefEffect( updateHeaderHeight, [] );
```
The header height doesn't vary so perhaps we should just hardcode it and eliminate this and the need for `Header` to forward a ref.

stokesman commented:
@talldan, I've rebased and reduced changes to what's relevant. I also reduced the added props to more tightly fit the present use case and added comments.

I think before anyone puts in the effort of a full code review the UX still needs vetted. So I'd like to bring up what has so far been the concern regarding this approach, mentioned by @kevin940726 and @critterverse 

https://github.com/WordPress/gutenberg/issues/32061#issuecomment-857894015

>I think it might be confusing to see the toolbar appear over other blocks after the selected block has scrolled out of view.

I've given that some thought but wager it would seldom cause confusion. I suppose a trial of it would be the only way to learn. We could also look at ways to indicate the selection is out-of-view. One quick idea:
![image](https://user-images.githubusercontent.com/9000376/121433757-47326e80-c931-11eb-98c2-5bffd62961fa.png). 

---- 
Toward the end of the screen recording in the description, I demonstrated dragging to reorder a block that was out of view. That was a nod toward looking out for such side effects. That particular one seems like it could be considered a feature by allowing an easier reordering task in some circumstances.

talldan commented:
@stokesman What do you think about moving the customizer title sticky changes into a separate PR? That part seems to add quite a bit of complexity and the performance needs improvement. I also think an option could be a patch to WordPress core for changing how it works.

If the PR just focuses on the block editor top bar and toolbar, that should hopefully make it a bit easier to iterate.

talldan commented:
BTW, I also had some code for reveal on scroll in this branch - https://github.com/WordPress/gutenberg/compare/trunk...try/customize-widget-revealing-topbar

That was an experiment to try the behavior on the editor top bar.

An approach I took there was to only change `top` when the scroll direction changes and let `position: sticky` do the rest of the work—set top to be on screen when scrolling down and off-screen when scrolling up. I think in this branch the `top` value is being modified constantly on every scroll event to match the scroll delta, which leads to the performance issues.

edit - actually looking at the code I seem to be misremembering, it seems to use the delta too. The performance difference might be more related to updating the React `style` prop, rather than modifying the DOM element's style directly.

stokesman commented:
>What do you think about moving the customizer title sticky changes into a separate PR?

A fine idea 👍  I'd been starting to think it myself. Certainly even if we find a performant solution it'd be intended as a crutch until core could be updated.

stokesman commented:
Removed the reimplementation of the Customizer’s sticky title. Number of lines changed greatly reduced 🎉 .

critterverse commented:
This is looking good with the most recent change to the Customizer title area 👍

>We could also look at ways to indicate the selection is out-of-view. 

Explored a couple of options for this and I think maybe the best option is potentially the most straightforward — dimming the toolbar after the selected block has scrolled out of view. The borders and icons are currently `Gutenberg Gray 900` (hex #1E1E1E) but could change to `Gutenberg Gray 600` (hex #949494) when dimmed:

https://user-images.githubusercontent.com/77359364/121610310-b38d9a80-ca23-11eb-844c-ad3645f39988.mov

---
Edit: Updating to include another exploration here as well since it might be helpful to get more opinions on this. In this version, the toolbar would anchor to the block as usual until it scrolls out of view, then collapse into a single-item toolbar (like a parent selector).

https://user-images.githubusercontent.com/77359364/121609825-b1770c00-ca22-11eb-8764-5bdece2bb86f.mov

kevin940726 commented:
https://github.com/WordPress/gutenberg/issues/32061#issuecomment-859273920

I wonder if this could work 🤔?

stokesman commented:
@critterverse, out of those explorations I tend to think the collapsing one would be the better of the two though probably a bit more effort. I'd also like to think we could allow the drag handle to still be available. Or maybe even a separate "scroll into view" button replace the ones that are hidden.

I've been thinking on it and coming around to prefer, at least for now, something like what @kevin940726 has suggested. I'd still like to see it done with the stickier toolbar as in this PR, so there's no overlap even while it transitions in and out.

critterverse commented:
>I'd still like to see it done with the stickier toolbar as in this PR, so there's no overlap even while it transitions in and out.

This sounds like a good plan but I'm having trouble picturing how the overlap can be avoided while fading — the toolbar and top bars are still going to intersect at some point before the toolbar can fade out, no?

Also added one more previously explored idea in #32061 where the block toolbar could go in front of the header bars rather than behind (like is happening currently with the footer controls at the bottom of the Customizer sidebar). 

I'm open to whichever approach we can get working better with the sticky editor bar, which is a nice usability improvement over what we have now!

stokesman committed:
Hide stickier popovers when their anchor is out of view

stokesman commented:
I've pushed an update to hide the block toolbars when their block is out of view. Instead of a fade, a transform is used since it was fun and may help reinforce why the toolbar is hiding. I'm not attached to that and fine with changing it. To try the fade while testing this PR the following can be pasted in to the styles in dev tools: 
```css
.components-popover.block-editor-block-list__block-popover .components-popover__content[data-away]{
    transition: opacity 0.1s linear;
}
.components-popover.block-editor-block-list__block-popover .components-popover__content[data-away].is-away {
  transform: none;
  opacity:0;
}
```

The screen recording in the description is updated.

>one more previously explored idea … where the block toolbar could go in front of the header bars rather than behind

That is definitely an improvement over trunk and happens to be where this PR started. I've put the first two commits here into a separate PR #32650.

critterverse commented:
Awesome, this was so helpful to be able to try out several different approaches at once!

>To try the fade while testing this PR the following can be pasted in to the styles in dev tools:

I really like how the version with the fade is working — props @stokesman and @kevin940726 for this elegant solution, it came together nicely with the sticky editor bar 🎉

stokesman committed:
Switch to fade transition and simplify

stokesman commented:
Thanks for testing it out @critterverse 🙇 . I've pushed an update to switch to the fade transition and that allowed a little simplification.

I've also simplified the logic determining if the selection is in view or not and that should ease code review. Which I think could be the next step here. Marking it ready.

noisysocks commented:
Thanks everyone for working on this. It's looking really good, but let's punt it to WordPress 5.8.1 as there's a few more rounds of UX/dev feedback required and we've only a limited amount of person hours available before the 5.8 code freeze on June 29.
